### PR TITLE
Model switching menu

### PR DESCRIPTION
**Pull request by Lukewh**
Implements https://github.com/juju/juju-gui/issues/2573

### CODE REVIEW

File: jujugui/static/gui/src/app/components/env-switcher/list/list.js [hunk 1]
```diff
@@ -120,29 +142,28 @@ YUI.add('env-list', function() {
 
     render: function() {
       const auth = this.props.authDetails;
-      let buttonRow;
+      let createNew;
       if (auth && auth.rootUserName) {
```
Comment (Lukewh) [inline jujugui/static/gui/src/app/components/env-switcher/list/list.js]: Not sure whether having the create new hidden behind auth is required? @hatched @frankban?

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/6929/

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/6930/

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/6931/

File: jujugui/static/gui/src/app/components/env-switcher/list/list.js [hunk 2]
```diff
@@ -47,18 +50,25 @@ YUI.add('env-list', function() {
...
-            No models available, click below to view your profile and create a
-            new model.
-          </li>
-        );
+        return false;
       }
       const auth = this.props.authDetails;
       const currentUser = auth ? auth.user : null;
-      return models.map(model => {
+      const modelsWithoutController = models.sort((a, b) => {
```
Comment (hatched) [inline jujugui/static/gui/src/app/components/env-switcher/list/list.js]: I see the code, but I have no idea why you're doing what you're doing here. Can you add some comments to this effect? And why would a model not have a controller?

File: jujugui/static/gui/src/app/components/env-switcher/list/list.js (hunk 1 above)
Comment (hatched) [inline jujugui/static/gui/src/app/components/env-switcher/list/list.js]: They can't create a model if they aren't logged in so yes.

Also, it should be disabled if they don't have the correct ACL for creating models.

Comment (hatched) [inline jujugui/static/gui/src/app/components/env-switcher/list/list.js]: https://github.com/juju/juju-gui/blob/develop/jujugui/static/gui/src/app/store/env/acl.js#L69

Comment (hatched) [inline jujugui/static/gui/src/app/components/env-switcher/list/list.js]: https://github.com/juju/juju-gui/blob/develop/jujugui/static/gui/src/app/app.js#L574

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/6937/

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/6938/

File: jujugui/static/gui/src/app/components/create-model-button/test-create-model-button.js [hunk 3]
```diff
@@ -47,22 +48,79 @@ describe('CreateModelButton', () => {
...
+        disabled={false}
+        changeState={sinon.stub()}
+        switchModel={sinon.stub()} />, true);
+    const output = component.getRenderOutput();
+    const instance = component.getMountedInstance();
+    instance._createNewModel();
+    assert.isTrue(action.called);
+  });
+
+  it('the passed action isn\'t called if button is disabled', () => {
```
Comment (hatched) [inline jujugui/static/gui/src/app/components/create-model-button/test-create-model-button.js]: Good tests thx

File: jujugui/static/gui/src/app/components/env-switcher/list/list.js [hunk 4]
```diff
@@ -54,12 +55,18 @@ YUI.add('env-list', function() {
...
+      // People shouldn't be editing the 'controller' model.
+      // They can still access it from their profile page.
+      const modelsWithoutController = models.filter(model => {
         return !model.isController &&
           model.name !== this.props.environmentName;
+      }).sort((a, b) => {
+        return b.lastConnection.getTime() - a.lastConnection.getTime();
       });
+      // If there is only one model left and it's the same name as the current
+      // environment - remove it.
```
Comment (hatched) [inline jujugui/static/gui/src/app/components/env-switcher/list/list.js]: Thanks for the comments

File: jujugui/static/gui/src/app/utils/component-test-utils.js [hunk 5]
```diff
@@ -347,33 +347,68 @@ var jsTestUtils = {
 
     @method specificDeepEqual
   */
-  specificDeepEqual: function(output, expected) {
```
Comment (hatched) [inline jujugui/static/gui/src/app/utils/component-test-utils.js]: I don't think you meant to commit this

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/6939/

**Comment by bac** [comment]
I am doing QA with one model of mine and three shared. This is the only scenario I exercised. I am logged in (clearly).

The model drop down allows me to go to other models.  But the 'Start a new model' is not clickable and therefore non-working.

QA-crawdaddy

**Comment by hatched** [comment]
The ACL method `canAddModels` returns `false` even if the user has the ability to do so on jaas.

**Comment by hatched** [comment]
After some investigation it turns out that the fields we need to check for this are returning the incorrect values. For the short term we can flag the ACL check on `gisf` so it's still respected with local Juju deploys.

**Comment by frankban** [comment]
Closing this as this is superceded by https://github.com/juju/juju-gui/pull/2647

**Pull request closed without merging**